src/node/mod.rs

```diff
@@ -0,0 +1,410 @@
+    pub fn dev_id(&self) -> dev_t {
+        self.dev
+    }
+
+    /// Returns the path of the open device if possible.
+    pub fn dev_path(&self) -> Option<PathBuf> {
+        node_path(self, self.ty).ok()
+    }
+
+    /// Returns the path of the specified node type matching the device, if available.
+    pub fn dev_path_with_type(&self, ty: NodeType) -> Option<PathBuf> {
+        node_path(self, ty).ok()
+    }
+
+    /// Returns a new node of the specified node type matching the device, if available.
+    pub fn node_with_type(&self, ty: NodeType) -> Option<Result<DrmNode, CreateDrmNodeError>> {
+        self.dev_path_with_type(ty).map(DrmNode::from_path)
+    }
+
+    /// Returns the major device number of the DRM device.
+    pub fn major(&self) -> u32 {
+        major(self.dev_id())
+    }
+
+    /// Returns the minor device number of the DRM device.
+    pub fn minor(&self) -> u32 {
+        minor(self.dev_id())
+    }
+
+    /// Returns whether the DRM device has render nodes.
+    pub fn has_render(&self) -> bool {
+        #[cfg(target_os = "linux")]
+        {
+            node_path(self, NodeType::Render).is_ok()
+        }
+
+        // TODO: More robust checks on non-linux.
+
+        #[cfg(target_os = "freebsd")]
+        {
+            false
+        }
+
+        #[cfg(not(any(target_os = "linux", target_os = "freebsd")))]
+        {
+            false
+        }
+    }
+}
+
+impl Display for DrmNode {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}{}", self.ty.minor_name_prefix(), minor(self.dev_id()))
+    }
+}
+
+/// A type of node
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub enum NodeType {
+    /// A primary node may be used to allocate buffers.
+    ///
+    /// If no other node is present, this may be used to post a buffer to an output with mode-setting.
+    Primary,
+
+    /// A control node may be used for mode-setting.
+    ///
+    /// This is almost never used since no DRM API for control nodes is available yet.
+    Control,
+
+    /// A render node may be used by a client to allocate buffers.
+    ///
+    /// Mode-setting is not possible with a render node.
+    Render,
+}
+
+impl NodeType {
+    /// Returns a string representing the prefix of a minor device's name.
+    ///
+    /// For example, on Linux with a primary node, the returned string would be `card`.
+    pub fn minor_name_prefix(&self) -> &str {
+        match self {
+            NodeType::Primary => PRIMARY_NAME,
+            NodeType::Control => CONTROL_NAME,
+            NodeType::Render => RENDER_NAME,
+        }
+    }
+}
+
+impl Display for NodeType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                NodeType::Primary => "Primary",
+                NodeType::Control => "Control",
+                NodeType::Render => "Render",
+            }
+        )
```

Comment:
This looks like the `Debug` implementation?

Perhaps use `f.write_str()` directly?

Reply:
I can see arguments going both ways tbh, this way the consistency of the implementation would be ensured at least.

But I've changed it to depend on the debug impl instead.